executor/indexer.py

```diff
@@ -0,0 +1,368 @@
+            buffer_matched_docs = deepcopy(docs[doc_idx].matches)
+            matched_docs = OrderedDict()
+            for m_info in zip(*matches):
+                idx, dist = m_info
+                if idx < 0 or self._metas['delete_marks'][idx]:
+                    continue
+                match_doc_id = self._metas['doc_ids'][idx]
+                match = self.get_doc(match_doc_id)
+                match.scores[self.metric] = dist
+
+                matched_docs[match.id] = match
+
+            # merge search results
+            for m in buffer_matched_docs:
+                matched_docs[m.id] = m
+
+            docs[doc_idx].matches = [
+                m
+                for _, m in sorted(
+                    matched_docs.items(),
+                    key=lambda item: item[1].scores[self.metric].value,
+                )
+            ][:top_k]
+
+    @requests(on='/update')
+    def update(self, docs: DocumentArray, parameters: Optional[Dict] = None, **kwargs):
+        """Update entries from the index by id
+        :param docs: the documents to update
+        :param parameters: parameters to the request
+        """
+
+        if (docs is None) or len(docs) == 0:
+            return
+
+        self._kv_db.update(docs)
+        sync = parameters.get('sync', True) if parameters else True
+        if sync:
+            self._buffer_indexer.extend(docs)
+
+    @requests(on='/delete')
+    def delete(self, parameters: Dict, **kwargs):
+        """Delete entries from the index by id
+        :param parameters: parameters to the request
+        """
+        deleted_ids = parameters.get('ids', [])
+        self._kv_db.delete(deleted_ids)
+
+        # delete from buffer_indexer
+        for doc_id in deleted_ids:
+            idx = self._metas['doc_id_to_offset'][doc_id]
+            self._metas['delete_marks'][idx] = 1
+
+            if doc_id in self._buffer_indexer:
+                del self._buffer_indexer[doc_id]
+
+    @requests(on='/sync')
+    def sync(self, **kwargs):
+        """Sync the data from the storage into the Faiss index"""
+        if self._vec_indexer:
+            self._vec_indexer.reset()
+        self._bloom = self._init_bloom()
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
+        self._vec_indexer = self._build_indexer(self._vec_indexer)
+        self._buffer_indexer.clear()
+
+    @requests(on='/clear')
+    def clear(self, **kwargs):
+        """Completely clear the storage and index."""
+        self._kv_db.clear()
+        if self._vec_indexer:
+            self._vec_indexer.reset()
+        self._buffer_indexer.clear()
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
+        self._bloom = self._init_bloom()
+
+    @requests(on='/status')
+    def status(self, **kwargs):
+        """Return the status of the indexer."""
+        status = Document(tags={'db_stat': self._kv_db.stat})
```

Comment:
Does this actually work? Doesn't it need to be appended as a match? Maybe add an integration test with a Flow

Reply:
you are right. We need an integration test to guarantee all endpoints work as expected. I created an issue https://github.com/jina-ai/executor-faissindexer/issues/9 to track this task.